src/subclass/types.rs

```diff
@@ -0,0 +1,421 @@
+use std::ptr;
+
+use translate::*;
+use wrapper::Wrapper;
+use {IsA, IsClassFor, Object, StaticType, Type};
+
+use super::object::ObjectImpl;
+
+/// A newly registered `glib::Type` that is currently still being initialized
+///
+/// This allows running additional type-setup functions, e.g. for implementing
+/// interfaces on the type
+#[derive(Debug, PartialEq, Eq)]
+pub struct InitializingType<T: ObjectSubclass>(Type, marker::PhantomData<T>);
+
+impl<T: ObjectSubclass> ops::Deref for InitializingType<T> {
+    type Target = Type;
+
+    fn deref(&self) -> &Type {
+        &self.0
+    }
+}
+
+/// Trait implemented by structs that implement a `GObject` C instance struct
+///
+/// The struct must be `#[repr(C)]` and have the parent type's instance struct
+/// as the first field.
+///
+/// See [`simple::InstanceStruct`] for a basic implementation of this that can
+/// be used most of the time and should only not be used if additional fields are
+/// required in the instance struct.
+///
+/// [`simple::InstanceStruct`]: ../simple/struct.InstanceStruct.html
+pub unsafe trait InstanceStruct: Sized + 'static {
+    /// Corresponding object subclass type for this instance struct
+    type Type: ObjectSubclass;
+
+    /// Returns the implementation for from this instance struct, that
+    /// is the implementor of [`ObjectImpl`] or subtraits.
+    ///
+    /// [`ObjectImpl`]: ../object/trait.ObjectImpl.html
+    fn get_impl(&self) -> &Self::Type {
+        unsafe {
+            let data = Self::Type::type_data();
+            let private_offset = data.as_ref().private_offset;
+            let ptr: *const u8 = self as *const _ as *const u8;
+            let priv_ptr = ptr.offset(private_offset);
+            let imp = priv_ptr as *const Option<Self::Type>;
+
+            (*imp).as_ref().expect("No private struct")
+        }
+    }
+
+    /// Returns the class struct for this specific instance
+    fn get_class(&self) -> &<Self::Type as ObjectSubclass>::Class {
+        unsafe { &**(self as *const _ as *const *const <Self::Type as ObjectSubclass>::Class) }
+    }
+}
+
+/// Trait implemented by structs that implement a `GObject` C class struct
+///
+/// The struct must be `#[repr(C)]` and have the parent type's class struct
+/// as the first field.
+///
+/// See [`simple::ClassStruct`] for a basic implementation of this that can
+/// be used most of the time and should only not be used if additional fields are
+/// required in the class struct, e.g. for declaring new virtual methods.
+///
+/// [`simple::ClassStruct`]: ../simple/struct.ClassStruct.html
+pub unsafe trait ClassStruct: Sized + 'static {
+    /// Corresponding object subclass type for this class struct
+    type Type: ObjectSubclass;
+
+    /// Override the vfuncs of all parent types
+    ///
+    /// This is automatically called during type initialization.
+    fn override_vfuncs(&mut self)
+    where
+        <<Self::Type as ObjectSubclass>::ParentType as Wrapper>::RustClassType:
```

Comment:
Is there a way to make this a bit less ugly? Using aliases and stuff could help I assume. The point would be to make this a bit easier to read...

Reply:
Unfortunately not possible because you can't define type aliases for associated type "evaluation" :(

I agree that this line is a bit hard to read. But fortunately this is nothing a user would ever have to explicitly use